syntax = "proto3";

package recon;

// Service definition for network reconnaissance
service ReconService {
  rpc Health (HealthRequest) returns (HealthResponse);
  rpc RunRecon (ReconRequest) returns (ReconResponse);
  rpc RunReconStream (ReconRequest) returns (stream ReconUpdate);
}

// Recon request message
message ReconRequest {
  repeated string domains = 1;
  bool silent_mode = 2;  // If true, skip direct HTTP requests to target
}

// DNS record
message DNSRecord {
  string type = 1;
  string value = 2;
  int32 ttl = 3;
}

// SSL certificate info from crt.sh
message SSLCertificate {
  string issuer = 1;
  string subject = 2;
  string not_before = 3;
  string not_after = 4;
  repeated string san_domains = 5;
}

// Security headers
message SecurityHeaders {
  string strict_transport_security = 1;
  string content_security_policy = 2;
  string x_frame_options = 3;
  string x_content_type_options = 4;
  string referrer_policy = 5;
  string permissions_policy = 6;
}

// WHOIS data
message WhoisData {
  string registrar = 1;
  string creation_date = 2;
  string expiration_date = 3;
  repeated string name_servers = 4;
  string registrant_org = 5;
  string status = 6;
}

// ASN information
message ASNInfo {
  string asn = 1;
  string organization = 2;
  repeated string ip_ranges = 3;
  string country = 4;

  // BGP routing data
  repeated string prefixes_v4 = 5;
  repeated string prefixes_v6 = 6;
  int32 prefix_count_v4 = 7;
  int32 prefix_count_v6 = 8;

  // Network relationships
  repeated string peers = 9;  // ASNs this network peers with
  repeated string upstreams = 10;  // Transit providers
  repeated string downstreams = 11;  // Customers

  // Peering info
  string peering_policy = 12;  // Open, Selective, Restrictive
  string network_type = 13;  // NSP, Content, Cable, Enterprise, etc.
  repeated string peering_facilities = 14;  // IX locations

  // Metadata
  string rir = 15;  // Regional Internet Registry (ARIN, RIPE, etc.)
  repeated string abuse_contacts = 16;
  string bgp_prefix = 17;  // The actual prefix the IP belongs to
}

// Domain recon result
message DomainRecon {
  string domain = 1;
  repeated DNSRecord dns_records = 2;
  repeated SSLCertificate ssl_certificates = 3;
  SecurityHeaders security_headers = 4;
  WhoisData whois = 5;
  ASNInfo asn_info = 6;
  repeated string subdomains = 7;
  string error = 8;
}

// Recon response
message ReconResponse {
  repeated DomainRecon results = 1;
}

// Recon update for streaming (can be a log or result)
message ReconUpdate {
  enum UpdateType {
    LOG = 0;
    RESULT = 1;
    COMPLETE = 2;
  }
  UpdateType type = 1;
  string message = 2;  // Log message
  DomainRecon result = 3;  // Final result for a domain
}

// Health check messages
message HealthRequest {}

message HealthResponse {
  string status = 1;  // "healthy" or "unhealthy"
  string message = 2;
}
